statistics/dml_update.go

```diff
@@ -0,0 +1,271 @@
+// limitations under the License.
+
+package statistics
+
+import (
+	"github.com/cznic/mathutil"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/codec"
+)
+
+// TableDelta stands for the changed stats for one table.
+// Note that we only update stats for primary key and indices.
+type TableDelta struct {
+	Delta     int64 // Delta is the delta count for a table.
+	Count     int64 // Count is the modify count for a table.
+	PKID      int64
+	PKDelta   *HistDelta
+	IdxDeltas map[int64]*HistDelta
+}
+
+// HistDelta stands for the changed stats for one histogram.
+type HistDelta struct {
+	HistVersion uint64 // HistVersion is the version of the histogram when the delta info is collected.
+	BktDeltas   map[int]*BktDelta
+}
+
+// BktDelta stands for the delta info for a bucket.
+type BktDelta struct {
+	Count int64
+	Bound *types.Datum // Bound is the new lower bound of the bucket. For the last bucket, it is also used to store the new upper bound of the bucket.
+}
+
+// Update updates the table delta according to the changed row.
+func (delta *TableDelta) Update(sc *stmtctx.StatementContext, h *Handle, tableInfo *model.TableInfo, row []types.Datum, deltaCount, count int64) {
+	delta.Delta += deltaCount
+	delta.Count += count
+	table := h.GetTableStats(tableInfo.ID)
+	// Update the primary key's histogram.
+	if tableInfo.PKIsHandle {
+		pk := tableInfo.GetPkColInfo()
+		c, ok := table.Columns[pk.ID]
+		if ok && pk.State == model.StatePublic && pk.Offset >= 0 && pk.Offset < len(row) {
+			if delta.PKDelta == nil {
+				delta.PKID = pk.ID
+				delta.PKDelta = &HistDelta{}
+			}
+			delta.PKDelta.update(sc, &c.Histogram, &row[pk.Offset], deltaCount)
+		}
+	}
+	// Update the indices' histogram.
+	for _, idxInfo := range tableInfo.Indices {
+		if idxInfo.State != model.StatePublic {
+			continue
+		}
+		idx, ok := table.Indices[idxInfo.ID]
+		if !ok {
+			continue
+		}
+		value := encodeValue(h.ctx.GetSessionVars().StmtCtx, idxInfo, row)
+		if value == nil {
+			continue
+		}
+		if delta.IdxDeltas == nil {
+			delta.IdxDeltas = make(map[int64]*HistDelta)
+		}
+		if delta.IdxDeltas[idxInfo.ID] == nil {
+			delta.IdxDeltas[idxInfo.ID] = &HistDelta{}
+		}
+		idxDelta := delta.IdxDeltas[idxInfo.ID]
+		idxDelta.update(sc, &idx.Histogram, value, deltaCount)
+	}
+}
+
+func encodeValue(sc *stmtctx.StatementContext, idxInfo *model.IndexInfo, row []types.Datum) *types.Datum {
+	values := make([]types.Datum, 0, len(idxInfo.Columns))
+	for _, col := range idxInfo.Columns {
+		if col.Offset < 0 || col.Offset >= len(row) {
+			return nil
+		}
+		values = append(values, row[col.Offset])
+	}
+	b, err := codec.EncodeKey(sc, nil, values...)
+	if err != nil {
+		terror.Log(err)
+		return nil
+	}
+	d := types.NewBytesDatum(b)
+	return &d
+}
+
+func (delta *BktDelta) update(sc *stmtctx.StatementContext, value *types.Datum, result int) {
```

Comment:
This update is only applied to bound, so this function's name doesn't fit. And the name of `result` is not good, maybe here should be a boolean value which stands for upper or lower bound.